src/stream/stream.rs

```diff
@@ -111,6 +127,48 @@ pub trait Stream {
             remaining: n,
         }
     }
+
+    /// Transforms a stream into a collection.
+    ///
+    /// `collect()` can take anything streamable, and turn it into a relevant
+    /// collection. This is one of the more powerful methods in the async
+    /// standard library, used in a variety of contexts.
+    ///
+    /// The most basic pattern in which `collect()` is used is to turn one
+    /// collection into another. You take a collection, call [`stream`] on it,
+    /// do a bunch of transformations, and then `collect()` at the end.
+    ///
+    /// Because `collect()` is so general, it can cause problems with type
+    /// inference. As such, `collect()` is one of the few times you'll see
+    /// the syntax affectionately known as the 'turbofish': `::<>`. This
+    /// helps the inference algorithm understand specifically which collection
+    /// you're trying to collect into.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # fn main() { async_std::task::block_on(async {
+    /// #
+    /// use async_std::prelude::*;
+    /// use async_std::stream;
+    ///
+    /// let s = stream::repeat(9u8).take(3);
+    /// let buf: Vec<u8> = s.collect().await;
+    ///
+    /// assert_eq!(buf, vec![9; 3]);
+    /// #
+    /// # }) }
+    /// ```
+    ///
+    /// [`stream`]: trait.Stream.html#tymethod.next
+    #[must_use = "if you really need to exhaust the iterator, consider `.for_each(drop)` instead (TODO)"]
+    fn collect<'a, B: FromStream<<Self as Stream>::Item>>(self) -> dyn_ret!('a, B)
```

Comment:
Any reason why the return type is `dyn Future` rather than `impl Future`? Shouldn't the `collect` method be an `async fn`?

Reply:
As far as I can see this is part of a trait rather than a concrete impl where `async fn` is allowed yet. Or am I missing something?

Reply:
Oh wait, I misread. I think you're right tho!

Reply:
@stjepang ahh, soooooooo -- this is the part I was talking about earlier; I tried *real* hard but couldn't get this to work using static futures. I gave up after a day of trying, and settled on obfuscated futures being "good enough". But if you think we really should have static futures here, maybe you could give it a shot?